### PR TITLE
✈️ Introduce Jetstream/Pytorch in TGI

### PR DESCRIPTION
**Pull request by tengomucho**
# What does this PR do?

This allows to use TGI with the `meta-llama/Llama-2-7b-hf` model using the Jetstream/Pytorch engine. 
This should be the starting point for a more complete integration in the future. It is not ready yet to replace the legacy implementation, in particular because:
- no other models have been tested, and some work is required for weights conversion;
- when I tried other Llama2 models it did not work, I still need to investigate the reason, this should follow next;
- further work should be done to simplify the coexistence of torch_xla and Jetstream/Pytorch. For now this feature is optional, in particular because the dependency installation is not that straightforward and execution of these two engines (Jetstream Pytorch and Pytorch/XLA) is exclusive.

## Before submitting
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: text-generation-inference/server/text_generation_server/jetstream_pt_support/generator.py [hunk 1]
```diff
@@ -0,0 +1,666 @@
...
+        tokens, true_length = pad_tokens(input_ids[0],
+                                         self.tokenizer.bos_token_id,
+                                         self.tokenizer.pad_token_id,
+                                         is_bos=True,
+                                         max_prefill_length=self.model.config.sequence_length,
+                                         jax_padding=True,
+                                         )
+        return tokens, true_length
+
+    def prefill(self, batch: Batch) -> Tuple[List[Generation], CachedBatch]:
```
[the remark this answers is not in the document]
Comment (tengomucho) [reply text-generation-inference/server/text_generation_server/jetstream_pt_support/generator.py:400]: Sure. `prefill` receives the request from the model server interface, you can see the code [here](https://github.com/huggingface/optimum-tpu/blob/5eb6cf3e696a20a5897cc3e1a9edf3b341a9b4e4/text-generation-inference/server/text_generation_server/server.py#L51).
The model server is called by the TGI router. You can see more information about TGI architecture [here](https://github.com/huggingface/text-generation-inference/blob/main/docs/source/architecture.md).
Let me know if you need any more specific detail!

**Review by allenwang28 — COMMENTED** [review]
Thanks Alvaro, great work! I took a first pass and left some fairly minor comments.

File: optimum/tpu/jetstream_pt_support.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    """Check if the necessary imports to use jetstream_pt are available.
+    """
+    try:
+        # For now Jetstream Pytorch is opt-in, it can be enabled with an ENV variable.
+        jetstream_pt_enabled = os.environ.get("JETSTREAM_PT", False) == "1"
+        if not jetstream_pt_enabled:
+            return False
+        # Torch XLA should not be imported before torch_xla2 to avoid conflicts.
+        if 'torch_xla2' not in sys.modules and 'torch_xla.core' in sys.modules:
+            return False
```
[the remark this answers is not in the document]
Comment (tengomucho) [reply optimum/tpu/jetstream_pt_support.py:15]: it's actually a little trickier than that: `torch_xla` cannot be imported _after_ `torch_xla` has been imported. I will add a warning.

File: text-generation-inference/server/text_generation_server/auto_generator.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
+from .generator_base import Generator
+from .jetstream_pt_support import check
```
[the remark this answers is not in the document]
Comment (tengomucho) [reply text-generation-inference/server/text_generation_server/auto_generator.py:2]: I rename it `model_can_use_jetstream_pt`.

File: text-generation-inference/server/text_generation_server/jetstream_pt_support/llama_model_exportable_hf.py
[the remark this answers is not in the document]
Comment (tengomucho) [reply text-generation-inference/server/text_generation_server/jetstream_pt_support/llama_model_exportable_hf.py:1]: That's what I tried to do at first, but if we want to support models as they are defined in `transformers`, the simplest way is to extract the model parameters from the config file. In the model definition in transformers, for Llama some of the original parameters (`hidden_dim`, `multiple_of` and `ffn_dim_multiplier`) were combined in the `intermediate_size` variable. I could not see a trivial way to go back to the original values, That is why I ended up re-implementing `FeedForward`, and as a consequence I ended up modifying the other classes that use that. If you think about a a way to get the original parameters back in a reliable way, then I can drop most of this and just use jetstream_pt's code.

File: text-generation-inference/server/text_generation_server/jetstream_pt_support/generator.py [hunk 4]
```diff
@@ -0,0 +1,666 @@
...
+    @property
+    def next_token(self) -> int:
+        return None if len(self._tokens) == 0 else self._tokens[-1]
+
+    @property
+    def empty(self) -> bool:
+        return len(self._tokens) == 0
+
+
+class TpuGeneratorJetStream(Generator):
```
[the remark this answers is not in the document]
Comment (tengomucho) [reply text-generation-inference/server/text_generation_server/jetstream_pt_support/generator.py:239]: Yes, I thought about it, and I agree with you that using only the engine and not the orchestrator means we will end up with different performance results. The reason why I did this was the API: the engine API is similar to TGI's model_server, while the orchestrator is not meant to interact via a Python API, but rather through gRPC, and its interface is more similar to the one in the TGI router. So interfacing the orchestrator with TGI would mean taking the TGI requests, re-encode them as requests for the jetstream orchestrator and forward them, then re-transcode the responses. So yes, at some point we might need to look at a way to integrate those, but it seems more complicated and I think we can do that later.

**Review by mfuntowicz — APPROVED** [review]

File: text-generation-inference/server/text_generation_server/jetstream_pt_support/logits_process.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+                The generation configuration to be used as base parametrization for the fused warper.
+
+        Returns:
+            a `FusedLogitsWarper` or None if neither top-k nor top-p are configured.
+        """
+        return cls(generation_config.temperature, generation_config.top_k, generation_config.top_p)
+
+    def __call__(self, logits: jnp.ndarray) -> Tuple[jnp.ndarray, jnp.ndarray]:
+        if self.temperature != 1.0:
+            logits = logits / self.temperature
```
Comment (miladm) [inline text-generation-inference/server/text_generation_server/jetstream_pt_support/logits_process.py:56]: qq: what happens if temp = 0?

Comment (tengomucho) [reply text-generation-inference/server/text_generation_server/jetstream_pt_support/logits_process.py:56]: Good question @miladm. In that the operation will give an array with [inf, -inf] values. The generation will still give some result, though probably not the one you would expect (in my case it was as if it was using greedy search).
BTW, you will have the same division in the Jetstream [sampling code](https://github.com/google/JetStream/blob/445f1aa8e857d0a09d72618e365daf80723bdf4c/jetstream/engine/sampling_utils.py#L40).

**Pull request merged**